### PR TITLE
Add support for `ALTER TABLE RENAME COLUMN`

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Resolves https://github.com/crate/crate/issues/10297

Add support for renaming columns that are top-level or sub-columns of partitioned/non-partitioned tables that are also part of constraints, partitioned by, clustered by, or generated columns.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
I see the current state as WIP which requires minor refinements/clean-ups and tests. Please look over `AlterTableRenameColumnAnalyzer` and `RenameColumnTask` which are the main components that enables this feature and see the dummy integration test, `test_alter_table_rename_column_dummy` for what it can currently achieve.

File: server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                genExpr,
+                Map.entry(SqlParser.createExpression(renameColumn.refToRename().column().sqlFqn()),
+                    SqlParser.createExpression(renameColumn.newName().sqlFqn())));
+            // TODO: make sure any quotings are preserved.
+            generatedColumns.put(e.getKey(), renamedGenExpr.toString());
+        }
+    }
+
+    /* Renames any columns within the expression. */
+    private static class ColumnRenameVisitor extends AstVisitor<Expression, Map.Entry<Expression, Expression>> {
```
Comment (jeeminso) [inline server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java:221]: This is needed to re-parse checks/generate expressions like - 
`abs(a + o['a']))`, `(o['a'] > a)` to safely rename columns in between. Any suggestions on this? If this seems ok, I will refine this further to more generic `ExpressionRewriter` and let `ColumnRenameVisitor` to extend it.


The only trouble right now is with quoting. `Expression.toString()` adds extra quotes. What comes to my mind in order to resolve quoting is to create another visitor that converts `Expression` to `ColumnIdent`(then to `String`) but I will see if I can avoid this.

File: server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        DocTableInfo tableInfo = (DocTableInfo) schemas.resolveTableInfo(
+            renameColumn.table().getName(),
+            Operation.ALTER,
+            txnCtx.sessionSettings().sessionUser(),
+            txnCtx.sessionSettings().searchPath());
+        var expressionAnalyzer = new ExpressionAnalyzer(
+            txnCtx,
+            nodeCtx,
+            paramTypeHints,
+            new NameFieldProvider(new DocTableRelation(tableInfo)),
```
Comment (seut) [inline server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java:68]: Please use a different (own) field provider instance so avoid using exceptions for the control flow. See e.g.  `TableElementsAnalyzer.resolveField` impl. as an example.

File: server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            Reference root = (Reference) expressionAnalyzer.convert(
+                SqlParser.createExpression(sourceRef.column().getRoot().sqlFqn()), expressionContext);
+            if (root.columnPolicy() == ColumnPolicy.IGNORED) {
+                throw new UnsupportedOperationException("Cannot rename sub-columns of ignored objects");
+            }
+        }
+        try {
+            expressionAnalyzer.convert(renameColumn.target(), expressionContext);
+        } catch (ColumnUnknownException e) {
+            // TODO: probably should not depend on exception
```
Comment (seut) [inline server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java:84]: See previous comment related to the field provider to avoid this.

**Review by seut — COMMENTED** [review]
Added more detailed comments/suggestions, overall looks good.

File: server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java (hunk 1 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java]: Thank you, this has been resolved. Please check the updates since the last review.

File: server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java (hunk 2 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java]: Thank you, please check the followup commits.

File: server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    private static void renameKeysFromPartitionedBy(Map<String, Object> mapping,
+                                                    AnalyzedAlterTableRenameColumn.RenameColumn renameColumn) {
+        List<List<String>> partitionedByCols = (List<List<String>>) mapping.get("partitioned_by");
+        for (var partitionedBy : partitionedByCols) {
+            // TODO: Use ColumnRenameVisitor
+            if (partitionedBy.get(0).equals(renameColumn.refToRename().column().fqn())) {
+                partitionedBy.removeFirst();
+                partitionedBy.addFirst(renameColumn.newName().fqn());
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java]: You meant within the loops right? I think the maps/lists modified here are copies already.

File: libs/sql-parser/src/main/java/io/crate/sql/tree/ExpressionRewriter.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.Map;
+
+/**
+ * This class can be extended to rewrite AST. Override the visit method for the target expression that needs to be
+ * re-written with a map entry having the key as the expression used to identify the expression to be re-written and the
+ * value as the expression to rewrite with. See {@link RenameColumnTask.ColumnRenameVisitor} as an example.
+ */
+
+public abstract class ExpressionRewriter extends DefaultTraversalVisitor<Expression, Map.Entry<Expression, Expression>> {
```
Comment (jeeminso) [inline libs/sql-parser/src/main/java/io/crate/sql/tree/ExpressionRewriter.java:32]: If this seems feasible, I will override the remaining 80ish methods.

**Comment by jeeminso** [comment]
Thank you for reviewing @seut , please have a second look. It should be complete except https://github.com/crate/crate/pull/15004#discussion_r1396485687.

File: server/src/test/java/io/crate/integrationtests/PgCatalogITest.java [hunk 6]
```diff
@@ -457,4 +457,12 @@ public void test_named_primary_key_constraint_names_are_visible() {
         execute("select conname from pg_catalog.pg_constraint where conname = 'c_aa'");
         assertThat(response).hasRows("c_aa");
     }
+
+    @Test
+    public void test_renamed_columns_are_reflected_on_pg_attribute() {
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/PgCatalogITest.java:462]: Could merge both `test_renamed_columns_are_reflected_on_pg_attribute` and `test_renamed_columns_are_reflected_on_information_schema` into a single `test_renamed_column_is_reflected_in_sys_tables`. Maybe even make it part of a one of the `AlterTableRenameColumnIntegrationTest` cases. It's all the same user scenario: Rename column must change the column name.

File: server/src/test/java/io/crate/execution/ddl/tables/RenameColumnTaskTest.java [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+                ObjectType.builder().setInnerType("o2", ObjectType.builder().setInnerType("x", DataTypes.INTEGER).build()).build(),
+                ColumnPolicy.DYNAMIC,
+                IndexType.PLAIN,
+                true,
+                false,
+                1,
+                1,
+                false,
+                null
+            );
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/execution/ddl/tables/RenameColumnTaskTest.java:102]: Could simplify the col-to-rename definitions by using `tbl.getReference(new ColumnIdent("o"))`. Would also avoid mistakenly using different properties in the SimpleReference

File: server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+    protected ClusterState execute(ClusterState currentState, RenameColumnRequest request) throws Exception {
+        var renameColumn = request.renameColumn();
+        var relationName = request.relationName();
+
+        var docTableInfoFactory = new DocTableInfoFactory(nodeContext);
+        DocTableInfo docTableInfo = docTableInfoFactory.create(relationName, currentState.metadata());
+        var txnCtx = CoordinatorTxnCtx.systemTransactionContext();
+        ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
+            txnCtx, nodeContext, ParamTypeHints.EMPTY, new AlterTableRenameColumnAnalyzer.FieldProviderResolvesUnknownColumns(docTableInfo), null);
+        ExpressionAnalysisContext expressionAnalysisContext = new ExpressionAnalysisContext(txnCtx.sessionSettings());
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java:89]: Could you try to re-implement this on top of https://github.com/crate/crate/pull/15055 by adding a `DocTableInfo.rename(from, to)` and re-using the `DocTableInfo.writeTo()` method?

File: server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+                    targetRef.column().getParent()));
+        }
+        if (!Objects.equals(sourceCol.getParent(), targetCol.getParent())) {
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ENGLISH,
+                    "When renaming sub-columns, parent names must be equal: %s, %s",
+                    sourceRef.column().getParent(),
+                    targetRef.column().getParent()));
+        }
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java:104]: Now come to think of it, I do not think so. I think allowing `rename column` to add dynamic objects could also be considered like how `add column` is allowed to do - `ALTER TABLE c ADD COLUMN o['a'] INT; `

File: docs/sql/statements/alter-table.rst [hunk 10]
```diff
@@ -210,6 +211,36 @@ It's possible to drop multiple columns at once.
...
+    It's allowed to rename a column:
+
+    - which is part of a :ref:`PRIMARY KEY <primary_key_constraint>`
+    - used in :ref:`CLUSTERED BY column <gloss-clustered-by-column>`
+    - used in :ref:`PARTITIONED BY <gloss-partitioned-by-column>`
+    - is referenced in a
+      :ref:`generated column <ddl-generated-columns-expressions>`
+    - is referenced in a
+      :ref:`table level constraint with other columns <check_constraint_multiple_cols>`
+
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply docs/sql/statements/alter-table.rst]: Removed, was thinking it would be reassuring. There should not be a limitation.

File: server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java:105]: >  I think allowing rename column to add dynamic objects could also be considered(to be discussed) 

I'd rather restrict that.
For the other scenarios: I think we can enable anything that just works with the generic "replace old columnIdent with new columnIdent" logic. 
If something requires additional custom logic, we should discuss if it's really needed

File: server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java (hunk 8 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java]: I think you can ignore that for now as it shouldn't cause test failures. As far as I can tell it's triggered by the templateUpgradeService - which triggers due to the version increment/change.

I'm not sure yet what's the right fix, but the templateUpgradeService probably shouldn't trigger validation of private settings

Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java]: Should be fixed with https://github.com/crate/crate/pull/15141

Comment (jeeminso) [reply server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java]: Thank you!

File: server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java]: Thank you, renaming columns that require depth changes lead to adjusting `innerType` of the parent object columns and also probably requires additional logic to take column policies into account, so I will move ahead with both scenarios restricted for now.

**Comment by jeeminso** [comment]
Thanks @mfussenegger could you review again?

Currently `DocTableInfo.renameColumn` does not fix `innerTypes` of `ObjectTypes` but it does not affect the functionality.
Also, `drop column` seems to be incomplete regarding `innerTypes` as well, e.g.:
```
cr> create table t (o object as (o2 object as (o3 object as (a int))));
cr> alter table t drop column o['o2']['o3']; -- innerType of o['o2'] is dropped but innerType of o still contains o3
```

It would be great if I can follow up to this `innerTypes` issue with respect to `rename column`, `drop column`, and `add column`.

File: server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        Metadata.Builder metadataBuilder = Metadata.builder(metadata);
+        Metadata newMetadata = changedTable
+            .writeTo(createMapperService, metadata, metadataBuilder)
+            .build();
+        // Ensure table can still be parsed
+        docTableInfoFactory.create(request.relationName(), newMetadata);
+        return ClusterState.builder(currentState)
+            .metadata(newMetadata)
+            .build();
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/ddl/tables/RenameColumnTask.java:66]: For a follow up: We could look into adding a `TableModifyTask` or something that's parameterized with something like `(request, table -> table)`. `AddColumnTask`, `DropColumnTask` and now `RenameColumnTask` all share the same approach: Get current table, modify it and create new ClusterState based on it.

File: server/src/main/java/io/crate/metadata/ColumnIdent.java [hunk 12]
```diff
@@ -196,6 +197,24 @@ public boolean isChildOf(ColumnIdent parent) {
...
+            return newName;
+        }
+        if (this.isChildOf(oldName)) {
+            var newPath = new ArrayList<>(newName.path());
+            newPath.addAll(this.path().subList(oldName.path().size(), this.path().size()));
+            return new ColumnIdent(newName.name(), newPath);
+        }
+        return this;
+    }
+
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/ColumnIdent.java:217]: I find the design of this strange. Renaming a `ColumnIdent` instance from an old name to a new name doesn't really make sense to me. The `newName` parameter is already the new name?

The function actually has a "rename-if-(sub)match" behavior but I don't understand why the path sub-replacement is needed.

It also seems a bit of a waste that we re-create _all_ `Reference` instances via their rename, despite only expecting to change the name of _one_ column.


Wouldn't it be more straightforward to re-use the `withReferenceIdent` - which already gives a Reference a new name - and use a `ref.column().equals(oldColumn) ? ref.withReferenceIdent(...) : ref` in the RenameTableTask?

**Review by mfussenegger — COMMENTED** [review]
> It would be great if I can follow up to this innerTypes issue with respect to rename column, drop column, and add column separately.

Sounds good to me.

File: server/src/main/java/io/crate/metadata/ColumnIdent.java (hunk 12 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/metadata/ColumnIdent.java:217]: > I find the design of this strange. Renaming a ColumnIdent instance from an old name to a new name doesn't really make sense to me. The newName parameter is already the new name?

If an object is renamed, the sub-columns need to be renamed as well so newName may need to be appended with child names.

> Wouldn't it be more straightforward to re-use the withReferenceIdent - which already gives a Reference a new name - and use a ref.column().equals(oldColumn) ? ref.withReferenceIdent(...) : ref in the RenameTableTask?

`Reference::rename` would propagate rename calls to its `Reference` fields which is more than what `withReferenceIdent` does. Would you like the the two methods to be merged? Also, could you explain why it should happen in `RenameColumnTask` instead of `DocTableInfo` ?

Comment (mfussenegger) [reply server/src/main/java/io/crate/metadata/ColumnIdent.java:217]: > If an object is renamed, the sub-columns need to be renamed as well so newName may need to be appended with child names.

Couldn't we make this a bit more concrete and explicit then?

> Reference::rename would propagate rename calls to its Reference fields which is more than what withReferenceIdent does

It seems to me as this is mixing a few different aspects all in the same - somewhat misleading - function.

- `Reference` can have a function which changes the name/ColumnIdent of it. This should of course update the name in _all_ of its relevant properties. If `withReferenceIdent` isn't doing that, that could be a bug

- _Other_ references might have had references to the renamed reference. They need an update too, but that's different than renaming a Reference

- The "if-match" aspect.





> Also, could you explain why it should happen in RenameColumnTask instead of DocTableInfo ?

I just mentioned `RenameColumnTask` because that's the entry point. Doing it in `DocTableInfo` is fine.

---

I kinda get that it fit's the current use-case, but the design still feels rather unusual. E.g. if you focus on the signature: `rename(ColumnIdent from, ColumnIdent newName) -> ColumnIdent` as instance method on a `ColumnIdent` - it's very hard to understand and confusing, because `newName` isn't the actual `newName` (for the object case you mentioned)

Comment (mfussenegger) [reply server/src/main/java/io/crate/metadata/ColumnIdent.java:217]: I'd also be interested what @seut or @matriv think about this. If it only bothers me, maybe it's fine to leave as is.

Comment (seut) [reply server/src/main/java/io/crate/metadata/ColumnIdent.java:217]: I agree with @mfussenegger about the mixed scope. I'd also prefer to separate them at e.g. `DocTableInfo.renameColumn` by something like:
```

Map<Reference, Reference> renamedReferences = references.values().stream()
  .filter(ref -> ident-is-equal-or-child-of-old-name)
  .collect(Collectors.toMap(ref, Reference.withReferenceIdent(ref.ident())));
```

And afterwards replace dependencies by using
```
 Function<Symbol, Symbol> updateRef = symbol -> RefReplacer.replaceRefs(symbol, ref -> renamedReferences.getOrDefault(ref, ref));
```

Comment (jeeminso) [reply server/src/main/java/io/crate/metadata/ColumnIdent.java]: Thank you all for your inputs, please check 2d076a9d5f644a40adf1798f2364d6ef1d60fe49 for update.

File: server/src/main/java/io/crate/metadata/doc/DocTableInfo.java [hunk 13]
```diff
@@ -745,6 +748,108 @@ public DocTableInfo dropColumns(List<DropColumn> columns) {
...
+        // where 1) is used to perform 2).
+
+        Map<ColumnIdent, Reference> oldNameToRenamedRefs = new HashMap<>();
+        for (var ref : references.values()) {
+            var r = toBeRenamed.test(ref.column()) ? ref.withReferenceIdent(
+                new ReferenceIdent(ref.ident().tableIdent(), ref.column().replacePath(newName))) : ref;
+            oldNameToRenamedRefs.put(
+                ref.column(), // old name
+                r); // renamed reference
+        }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/doc/DocTableInfo.java:783]:
```suggestion
        for (var ref : references.values()) {
            ColumnIdent column = ref.column();
            if (toBeRenamed.test(column)) {
                var renamedRef = ref.withReferenceIdent(
                    new ReferenceIdent(ident, ref.column().replacePath(newName)));

                oldNameToRenamedRefs.put(column, renamedRef);
            }
        }
```

Looks like it's not necessary to add the references which aren't renamed. The lookups use `getOrDefault()`
And no need to add a comment with a better name if you can rename the variable itself.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetadata.java [hunk 14]
```diff
@@ -130,6 +130,8 @@ public Settings settings() {
         return this.settings;
     }
 
+    
+
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateMetadata.java:134]: Unrelated style change - could revert to avoid the noise

File: server/src/test/java/io/crate/analyze/AlterTableRenameColumnAnalyzerTest.java [hunk 15]
```diff
@@ -0,0 +1,154 @@
+package io.crate.analyze;
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/analyze/AlterTableRenameColumnAnalyzerTest.java:1]: Is checkstyle not complaining about missing header?

File: server/src/test/java/io/crate/integrationtests/AlterTableRenameColumnIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+import org.junit.Test;
+
+import io.crate.exceptions.SQLParseException;
+import io.crate.testing.Asserts;
+import io.crate.testing.UseJdbc;
+
+public class AlterTableRenameColumnIntegrationTest extends IntegTestCase {
+
+    @Test
+    public void test_rename_primary_key_columns() {
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/AlterTableRenameColumnIntegrationTest.java:38]: Do we really need to cover all scenarios with both the RenameTaskTest and integration test?

I think it would be safe to reduce this to one integration test - to ensure it's wired up correctly and to ensure the change is reflected in show create table and the system tables - and to rely on the `RenameTaskTest` for the rest?

File: server/src/main/java/io/crate/metadata/doc/DocTableInfo.java [hunk 17]
```diff
@@ -745,6 +748,108 @@ public DocTableInfo dropColumns(List<DropColumn> columns) {
...
+        };
+
+        Function<IndexReference, IndexReference> renameIndexRefs = idxRef -> {
+            var dependenciesRenamed = idxRef.updateColumns(idxRef.columns().stream()
+                    .map(r -> oldNameToRenamedRefs.getOrDefault(r.column(), r)).toList());
+            if (toBeRenamed.test(idxRef.column())) {
+                return (IndexReference) dependenciesRenamed.withReferenceIdent(
+                    new ReferenceIdent(idxRef.ident().tableIdent(), idxRef.column().replacePath(newName)));
+            }
+            return dependenciesRenamed;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/doc/DocTableInfo.java:802]: `dependenciesRenamed` sounds like a collection due to the use of plural. Could use either `IndexReference dependenciesRenamed` or just `var updatedRef`

File: server/src/main/java/io/crate/metadata/doc/DocTableInfo.java [hunk 18]
```diff
@@ -745,6 +748,108 @@ public DocTableInfo dropColumns(List<DropColumn> columns) {
...
+                    genRef.reference(), // already renamed in step 1)
+                    RefReplacer.replaceRefs(genRef.generatedExpression(), r -> oldNameToRenamedRefs.getOrDefault(r.column(), r))
+                );
+            }
+            return ref;
+        };
+
+        Function<IndexReference, IndexReference> renameIndexRefs = idxRef -> {
+            var dependenciesRenamed = idxRef.updateColumns(idxRef.columns().stream()
+                    .map(r -> oldNameToRenamedRefs.getOrDefault(r.column(), r)).toList());
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/doc/DocTableInfo.java:797]: No need to change, but just fyi: `Lists2.map` is usually faster & shorter than the stream API if you just need to do a map

File: server/src/main/java/io/crate/metadata/ColumnIdent.java [hunk 19]
```diff
@@ -196,6 +197,21 @@ public boolean isChildOf(ColumnIdent parent) {
...
 
+    /**
+     * Replaces the parent name of this column specified by the 'newPath' e.g.::
+     *   'a.b.c'.replacePath('a.x') -> 'a.x.c' : parent 'a.b' renamed to 'a.x' such that its child 'a.b.c' is renamed to 'a.x.c'
+     *   'a.b.c'.replacePath('x') -> 'x.b.c' : parent 'a' renamed to 'x' such that its child 'a.b.c' is renamed to 'x.b.c'
+     * This method also renames self e.g.::
+     *   'a'.replacePath('b') -> 'b'
+     */
+    @NotNull
+    public ColumnIdent replacePath(@NotNull ColumnIdent newPath) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/ColumnIdent.java:208]: Not sure about this - but maybe `replaceHead` or `replacePrefix` would make this clearer?
It's replacing the starting path segment with the new prefix, and keeps any extra tail it had.

**Review by mfussenegger — APPROVED** [review]
I left a couple more comments about some implementation details. Overall looks good.

File: server/src/test/java/io/crate/integrationtests/AlterTableRenameColumnIntegrationTest.java (hunk 16 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/integrationtests/AlterTableRenameColumnIntegrationTest.java:38]: Thank you, reduced to a single test and move it to `AlterTableIntegrationTest`.

File: server/src/test/java/io/crate/analyze/AlterTableRenameColumnAnalyzerTest.java (hunk 15 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/analyze/AlterTableRenameColumnAnalyzerTest.java:22]: We no longer get the header auto appended when creating a new file and also check style does not check anymore. Need to fix: https://github.com/crate/crate/blob/6f9d72ac44d36cbfff50e8a0813daff7369f600c/server/src/test/java/io/crate/execution/ddl/tables/RenameColumnTaskTest.java#L1
